src/rust/ide/src/controller/upload.rs

```diff
@@ -0,0 +1,625 @@
+}
+
+
+
+// =========================
+// === FileUploadProcess ===
+// =========================
+
+/// Information about file-to-upload.
+#[allow(missing_docs)]
+#[derive(Clone,Debug)]
+pub struct FileToUpload<DataProvider> {
+    pub name : String,
+    pub size : u64,
+    pub data : DataProvider
+}
+
+/// The handler of uploading a given file to the specific location using the Language Server's file
+/// API.
+#[derive(Clone,Debug)]
+pub struct FileUploadProcess<DataProvider> {
+    logger          : Logger,
+    bin_connection  : Rc<binary::Connection>,
+    json_connection : Rc<language_server::Connection>,
+    file            : FileToUpload<DataProvider>,
+    remote_path     : Path,
+    bytes_uploaded  : u64,
+    checksum        : sha3::Sha3_224,
+}
+
+impl<DP:DataProvider> FileUploadProcess<DP> {
+    /// Constructor.
+    pub fn new
+    ( parent          : impl AnyLogger
+    , file            : FileToUpload<DP>
+    , bin_connection  : Rc<binary::Connection>
+    , json_connection : Rc<language_server::Connection>
+    , remote_path : Path
+    ) -> Self {
+        let logger         = Logger::sub(parent,"FileUploadProcess");
+        let bytes_uploaded = 0;
+        let checksum       = sha3::Sha3_224::new();
+        Self {logger,bin_connection,json_connection,file,remote_path,bytes_uploaded,checksum}
+    }
+
+    /// Upload next chunk. Returns true if all data has been uploaded.
+    ///
+    /// After uploading, the checksum of the uploaded file is compared with the file content digest,
+    /// and an error is returned if they do not match.
+    pub async fn upload_chunk(&mut self) -> FallibleResult<bool> {
+        match self.file.data.next_chunk().await {
+            Ok(Some(data)) => {
+                debug!(self.logger, "Received chunk of {self.file.name} of size {data.len()} \
+                    uploading to {self.remote_path:?}: {data:?}");
+                let offset = self.bytes_uploaded;
+                self.bin_connection.write_bytes(&self.remote_path,offset,false,&data).await?;
+                self.checksum.input(&data);
+                self.bytes_uploaded += data.len() as u64;
+                Ok(false)
+            },
+            Ok(None) => {
+                if self.bytes_uploaded != self.file.size {
+                    error!(self.logger, "The promised file size ({self.file.size}) and uploaded \
+                        data length ({self.bytes_uploaded}) do not match. Leaving as much data as \
+                        received.");
+                    self.bytes_uploaded = self.file.size;
+                }
+                self.check_checksum().await?;
+                Ok(true)
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    async fn check_checksum(&mut self) -> FallibleResult {
+        let remote = self.json_connection.file_checksum(&self.remote_path).await?.checksum;
+        let local  = std::mem::take(&mut self.checksum).into();
```

Comment:
The checksum is taken, because the `result` method on sha3::Sha3_224 takes `self`. I could only call result_reset instead, but still the errors will be there. And I opt to avoid cloning this structure.

The checksum should be finalized only once per upload; I added a note in `upload_chunk` method about subsequent calls after finishing uploading.